File: Little Artist/Services/NotificationService.swift
//
//  NotificationService.swift
//  Little Artist
//
//  Schedules local notifications for inactivity reminders
//  and "On This Day" artwork memories.
//

import UserNotifications
import SwiftData

/// Handles local notification scheduling for artwork reminders and memories.
///
/// Provides two notification types:
/// - **Inactivity reminder** — fires 14 days after the most recent artwork capture.
/// - **On This Day** — fires at 9 AM each day for artworks created on the same
///   month and day in a previous year.
enum NotificationService {

    // MARK: - Identifiers

    private static let inactivityIdentifier = "com.littleartist.inactivityReminder"
    private static let onThisDayPrefix = "com.littleartist.onThisDay."
    private static let sharedChangePrefix = "com.littleartist.sharedChange."

    /// Number of days of inactivity before sending a reminder.
    private static let inactivityDays: Int = 14

    // MARK: - Permission

    /// Requests notification authorization from the user.
    ///
    /// - Returns: `true` if the user granted permission.
    @discardableResult
    static func requestPermission() async -> Bool {
        do {
            let granted = try await UNUserNotificationCenter.current()
                .requestAuthorization(options: [.alert, .badge, .sound])
            return granted
        } catch {
            return false
        }
    }

    // MARK: - Schedule All

    /// Cancels existing notifications and schedules new ones based on the
    /// provided artworks.
    ///
    /// - Parameter artworks: All artworks currently stored in the app.
    static func scheduleAll(artworks: [Artwork]) {
        cancelAll()

        // Inactivity reminder based on the most recent artwork date
        let mostRecentDate = artworks.map(\.createdAt).max()
        scheduleInactivityReminder(lastArtworkDate: mostRecentDate ?? Date.now)

        // On This Day memories
        scheduleOnThisDayNotifications(artworks: artworks)
    }

    // MARK: - Cancel All

    /// Removes all pending notification requests scheduled by this service.
    static func cancelAll() {
        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
    }

    // MARK: - Inactivity Reminder

    /// Schedules a notification that fires 14 days after `lastArtworkDate`.
    ///
    /// If the trigger date is already in the past (i.e., it has been more than
    /// 14 days since the last artwork), the notification fires in 1 minute.
    ///
    /// - Parameter lastArtworkDate: The `createdAt` date of the most recent artwork.
    static func scheduleInactivityReminder(lastArtworkDate: Date) {
        let content = UNMutableNotificationContent()
        content.title = "Time to Create!"
        content.body = "You haven't captured any artwork recently! Time to add some new creations."
        content.sound = .default

        let triggerDate = Calendar.current.date(
            byAdding: .day,
            value: inactivityDays,
            to: lastArtworkDate
        ) ?? lastArtworkDate

        let trigger: UNNotificationTrigger

        if triggerDate > Date.now {
            let interval = triggerDate.timeIntervalSince(Date.now)
            trigger = UNTimeIntervalNotificationTrigger(timeInterval: interval, repeats: false)
        } else {
            // Already overdue — fire in 60 seconds
            trigger = UNTimeIntervalNotificationTrigger(timeInterval: 60, repeats: false)
        }

        let request = UNNotificationRequest(
            identifier: inactivityIdentifier,
            content: content,
            trigger: trigger
        )

        UNUserNotificationCenter.current().add(request)
    }

    // MARK: - On This Day

    /// Schedules a daily 9 AM notification for each artwork whose creation
    /// date shares the current month and day in a previous year.
    ///
    /// - Parameter artworks: All artworks to evaluate for matching dates.
    static func scheduleOnThisDayNotifications(artworks: [Artwork]) {
        let calendar = Calendar.current
        let today = calendar.dateComponents([.month, .day], from: Date.now)
        guard let todayMonth = today.month, let todayDay = today.day else { return }

        for artwork in artworks {
            let components = calendar.dateComponents([.year, .month, .day], from: artwork.createdAt)
            guard
                components.month == todayMonth,
                components.day == todayDay,
                let artworkYear = components.year
            else { continue }

            let yearsAgo = calendar.component(.year, from: Date.now) - artworkYear
            guard yearsAgo >= 1 else { continue }

            let childName = artwork.child?.name ?? "Your little artist"
            let artworkTitle = artwork.title.isEmpty ? "an artwork" : "'\(artwork.title)'"
            let yearLabel = yearsAgo == 1 ? "1 year" : "\(yearsAgo) years"

            let content = UNMutableNotificationContent()
            content.title = "On This Day"
            content.body = "On this day \(yearLabel) ago, \(childName) created \(artworkTitle)!"
            content.sound = .default

            // Schedule for 9 AM today
            var triggerComponents = DateComponents()
            triggerComponents.hour = 9
            triggerComponents.minute = 0

            let trigger = UNCalendarNotificationTrigger(
                dateMatching: triggerComponents,
                repeats: false
            )

            let identifier = "\(onThisDayPrefix)\(artwork.persistentModelID.hashValue)"
            let request = UNNotificationRequest(
                identifier: identifier,
                content: content,
                trigger: trigger
            )

            UNUserNotificationCenter.current().add(request)
        }
    }

    // MARK: - Shared Artwork Changes

    /// Sends a local notification when shared artworks are added or updated
    /// by another user.
    ///
    /// Uses a 3-second delay to batch rapid successive syncs. The identifier
    /// is based on the child name hash so successive notifications for the
    /// same child replace (not stack) each other.
    ///
    /// - Parameters:
    ///   - childName: The name of the shared child profile.
    ///   - insertedCount: Number of newly added artworks.
    ///   - updatedCount: Number of updated artworks.
    static func notifySharedArtworkChanges(childName: String, insertedCount: Int, updatedCount: Int) {
        guard UserDefaults.standard.bool(forKey: "notificationsEnabled") else { return }
        guard insertedCount > 0 || updatedCount > 0 else { return }

        let content = UNMutableNotificationContent()
        content.sound = .default

        if insertedCount > 0 && updatedCount == 0 {
            let artworkWord = insertedCount == 1 ? "artwork" : "artworks"
            content.title = "New Artwork for \(childName)"
            content.body = "\(insertedCount) new \(artworkWord) added to \(childName)'s gallery."
        } else if updatedCount > 0 && insertedCount == 0 {
            content.title = "\(childName)'s Gallery Updated"
            content.body = "\(updatedCount) \(updatedCount == 1 ? "artwork" : "artworks") updated."
        } else {
            content.title = "\(childName)'s Gallery Updated"
            content.body = "\(insertedCount) new, \(updatedCount) updated."
        }

        // 3-second delay batches rapid changes
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 3, repeats: false)

        // Use child name hash so successive syncs replace, not stack
        let identifier = "\(sharedChangePrefix)\(childName.hashValue)"
        let request = UNNotificationRequest(
            identifier: identifier,
            content: content,
            trigger: trigger
        )

        UNUserNotificationCenter.current().add(request)
    }
}

File: little_artist_flutter/lib/services/notification_service.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:timezone/data/latest_all.dart' as tz;
import 'package:timezone/timezone.dart' as tz;

import 'package:little_artist/models/database.dart';

/// Handles local notification scheduling for artwork reminders and memories.
///
/// Provides two notification types:
/// - **Inactivity reminder** -- fires 14 days after the most recent artwork.
/// - **On This Day** -- fires at 9 AM each day for artworks created on the
///   same month and day in a previous year.
abstract final class NotificationService {
  // Identifiers
  static const _inactivityId = 0;
  static const _onThisDayIdBase = 1000;

  /// Number of days of inactivity before sending a reminder.
  static const _inactivityDays = 14;

  static final FlutterLocalNotificationsPlugin _plugin =
      FlutterLocalNotificationsPlugin();

  static bool _initialised = false;

  // ---------------------------------------------------------------------------
  // Initialisation
  // ---------------------------------------------------------------------------

  /// Initialises the notifications plugin and timezone data.
  ///
  /// Safe to call multiple times; only runs once.
  static Future<void> _ensureInitialised() async {
    if (_initialised) return;

    tz.initializeTimeZones();
    final localName = tz.local.name;
    tz.setLocalLocation(tz.getLocation(localName));

    const androidSettings =
        AndroidInitializationSettings('@mipmap/ic_launcher');
    const iosSettings = DarwinInitializationSettings(
      requestAlertPermission: false,
      requestBadgePermission: false,
      requestSoundPermission: false,
    );

    const settings = InitializationSettings(
      android: androidSettings,
      iOS: iosSettings,
    );

    await _plugin.initialize(settings);
    _initialised = true;
  }

  // ---------------------------------------------------------------------------
  // Permission
  // ---------------------------------------------------------------------------

  /// Requests notification authorization from the user.
  ///
  /// Returns `true` if the user granted permission.
  static Future<bool> requestPermission() async {
    await _ensureInitialised();

    // iOS
    final iosPlugin = _plugin.resolvePlatformSpecificImplementation<
        IOSFlutterLocalNotificationsPlugin>();
    if (iosPlugin != null) {
      final granted = await iosPlugin.requestPermissions(
        alert: true,
        badge: true,
        sound: true,
      );
      return granted ?? false;
    }

    // Android
    final androidPlugin = _plugin.resolvePlatformSpecificImplementation<
        AndroidFlutterLocalNotificationsPlugin>();
    if (androidPlugin != null) {
      final granted = await androidPlugin.requestNotificationsPermission();
      return granted ?? false;
    }

    return false;
  }

  // ---------------------------------------------------------------------------
  // Schedule all
  // ---------------------------------------------------------------------------

  /// Cancels existing notifications and schedules new ones based on the
  /// provided artworks.
  static Future<void> scheduleAll(List<Artwork> artworks) async {
    await _ensureInitialised();
    await cancelAll();

    // Inactivity reminder based on the most recent artwork date
    DateTime? mostRecentDate;
    for (final artwork in artworks) {
      if (mostRecentDate == null ||
          artwork.createdAt.isAfter(mostRecentDate)) {
        mostRecentDate = artwork.createdAt;
      }
    }
    await scheduleInactivityReminder(mostRecentDate ?? DateTime.now());

    // On This Day memories
    await scheduleOnThisDayNotifications(artworks);
  }

  // ---------------------------------------------------------------------------
  // Cancel all
  // ---------------------------------------------------------------------------

  /// Removes all pending notification requests scheduled by this service.
  static Future<void> cancelAll() async {
    await _ensureInitialised();
    await _plugin.cancelAll();
  }

  // ---------------------------------------------------------------------------
  // Inactivity reminder
  // ---------------------------------------------------------------------------

  /// Schedules a notification that fires [_inactivityDays] days after
  /// [lastArtworkDate].
  ///
  /// If the trigger date is already in the past, the notification fires
  /// in 1 minute.
  static Future<void> scheduleInactivityReminder(
    DateTime lastArtworkDate,
  ) async {
    await _ensureInitialised();

    const details = NotificationDetails(
      android: AndroidNotificationDetails(
        'reminders',
        'Reminders',
        channelDescription: 'Artwork capture reminders',
        importance: Importance.high,
        priority: Priority.defaultPriority,
      ),
      iOS: DarwinNotificationDetails(),
    );

    final triggerDate = lastArtworkDate.add(
      const Duration(days: _inactivityDays),
    );
    final now = DateTime.now();

    if (triggerDate.isAfter(now)) {
      final scheduledDate = tz.TZDateTime.from(triggerDate, tz.local);
      await _plugin.zonedSchedule(
        _inactivityId,
        'Time to Create!',
        "You haven't captured any artwork recently! Time to add some new creations.",
        scheduledDate,
        details,
        androidScheduleMode: AndroidScheduleMode.inexactAllowWhileIdle,
        uiLocalNotificationDateInterpretation:
            UILocalNotificationDateInterpretation.absoluteTime,
      );
    } else {
      // Already overdue -- fire in 60 seconds
      final soon = tz.TZDateTime.now(tz.local).add(
        const Duration(seconds: 60),
      );
      await _plugin.zonedSchedule(
        _inactivityId,
        'Time to Create!',
        "You haven't captured any artwork recently! Time to add some new creations.",
        soon,
        details,
        androidScheduleMode: AndroidScheduleMode.inexactAllowWhileIdle,
        uiLocalNotificationDateInterpretation:
            UILocalNotificationDateInterpretation.absoluteTime,
      );
    }
  }

  // ---------------------------------------------------------------------------
  // On This Day
  // ---------------------------------------------------------------------------

  /// Schedules a daily 9 AM notification for each artwork whose creation
  /// date shares the current month and day in a previous year.
  static Future<void> scheduleOnThisDayNotifications(
    List<Artwork> artworks,
  ) async {
    await _ensureInitialised();

    final now = DateTime.now();
    final todayMonth = now.month;
    final todayDay = now.day;
    int notificationId = _onThisDayIdBase;

    for (final artwork in artworks) {
      final createdAt = artwork.createdAt;
      if (createdAt.month != todayMonth || createdAt.day != todayDay) continue;

      final yearsAgo = now.year - createdAt.year;
      if (yearsAgo < 1) continue;

      final yearLabel = yearsAgo == 1 ? '1 year' : '$yearsAgo years';
      final artworkTitle =
          artwork.title.isEmpty ? 'an artwork' : "'${artwork.title}'";

      const details = NotificationDetails(
        android: AndroidNotificationDetails(
          'reminders',
          'Reminders',
          channelDescription: 'Artwork capture reminders',
          importance: Importance.high,
          priority: Priority.defaultPriority,
        ),
        iOS: DarwinNotificationDetails(),
      );

      // Schedule for 9 AM today
      final scheduledDate = tz.TZDateTime(
        tz.local,
        now.year,
        now.month,
        now.day,
        9,
        0,
      );

      // Only schedule if 9 AM hasn't passed yet
      if (scheduledDate.isAfter(tz.TZDateTime.now(tz.local))) {
        try {
          await _plugin.zonedSchedule(
            notificationId,
            'On This Day',
            'On this day $yearLabel ago, your little artist created $artworkTitle!',
            scheduledDate,
            details,
            androidScheduleMode: AndroidScheduleMode.inexactAllowWhileIdle,
            uiLocalNotificationDateInterpretation:
                UILocalNotificationDateInterpretation.absoluteTime,
          );
        } catch (e) {
          debugPrint('Failed to schedule On This Day notification: $e');
        }
      }

      notificationId++;
    }
  }
}
